### ISSUE TITLE
Unexpected type casting from `torch.bfloat16` to `torch.float32`

### ISSUE THREAD

**Issue by pietrolesci** [huggingface/transformers#29349]
Hi there,

I am loading a `BertForSequenceClassification` model in (true) `bfloat16` precision. I noticed that when I manually implement the steps in the [forward pass](https://github.com/huggingface/transformers/blob/a0857740c0e6127485c11476650314df3accc2b6/src/transformers/models/bert/modeling_bert.py#L1543) (i.e., get the pooler output, apply dropout, and pass it through the classifier head) the logits type is, as expected, `bfloat16`. However, if I simply pass a batch in the forward method and then get the logits, their type is now `float32`. In code,

```python
inp_ids, att_mask = batch["input_ids"], batch["attention_mask"]

model(input_ids=inp_ids, attention_mask=att_mask).logits.dtype
#>>> torch.float32

pooled_out = model.bert(input_ids=inp_ids, attention_mask=att_mask).pooler_output
pooled_out = model.dropout(pooled_out)
model.classifier(pooled_out).dtype
#>>> torch.bfloat16
```

This seems unexpected. I think the issue is due to the implementation of the `SequenceClassifierOutput` https://github.com/huggingface/transformers/blob/49204c1d37b807def930fe45f5f84abc370a7200/src/transformers/modeling_outputs.py#L1010-L1013

**Comment by ArthurZucker** [comment]
Hey! I doubt the type hint influence the casting of the inputs. Also many things can play into this, the output is better in float32 if you are doing softmax afterward, see Llama for example

**Comment by ArthurZucker** [comment]
Alright, closing then! It's the same for Llama models, unless you manually reproduce every single operation! 🤗
